libutils/sequence.c

```diff
@@ -449,3 +451,114 @@ int SeqStringLength(Seq *seq)
 
     return total_length;
 }
+
+// TODO: These static helper functions could be (re)moved
+static bool HasNulByte(const char *str, size_t n)
+{
+    for (int i = 0; i < n; ++i)
+    {
+        if (str[i] == '\0')
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+static long GetLengthPrefix(const char *data)
+{
+    if (HasNulByte(data, 10))
+    {
+        return -1;
+    }
+
+    if (!isdigit(data[0]))
+    {
+        return -1;
+    }
+
+    if (data[SEQ_PREFIX_LEN-1] != ' ')
+    {
+        return -1;
+    }
+
+    // NOTE: This uses unsigned long because HPUX sscanf doesn't support %zu
+    unsigned long length;
+    int ret = sscanf(data, "%lu", &length);
+    if (ret != 1)
+    {
+        // incorrect number of items scanned
+        return -1;
+    }
+
+    return (long)length;
+}
+
+static char *ValidDuplicate(const char *src, long n)
+{
+    assert(src != NULL);
+    assert(n >= 0);
+    char *dst = xcalloc(n+1, sizeof(char));
+
+    long len = StringCopy(dst, src, n);
+    if (len < n) // string was too short
+    {
+        free(dst);
+        return NULL;
+    }
+
+    return dst;
+}
+
+char *SeqStringSerialize(Seq *seq)
+{
+    assert(seq != NULL);
+    size_t length = SeqLength(seq);
+    Writer *w = StringWriter();
+
+    for (int i = 0; i < length; ++i)
+    {
+        const char *s = SeqAt(seq, i);
+        const unsigned long str_length = strlen(s);
+        WriterWriteF(w, "%-" TOSTRING(SEQ_PREFIX_LEN) "lu%s\n", str_length, s);
+    }
+
+    return StringWriterClose(w);
+}
+
+Seq *SeqStringDeserialize(const char *const serialized)
+{
+    assert(serialized != NULL);
+    assert(SEQ_PREFIX_LEN > 0);
+
+    Seq *seq = SeqNew(512, free);
+
+    const char *src = serialized;
+    while (src[0] != '\0')
+    {
+        // Read length prefix first
+        long length = GetLengthPrefix(src);
+
+        // Advance the src pointer
+        src += SEQ_PREFIX_LEN;
+
+        char *new_str;
+
+        // Do validation and duplication in one pass
+        if (length < 0
+            || NULL == (new_str = ValidDuplicate(src, length))
+            || src[-1] != ' '
+            || src[length] != '\n')
```

Comment:
I know this version of the code is not focusing on performance, but still these checks could be reordered to do the most expensive one last.

Reply:
Oh, I see.